src/rollback_to_stable/rts_btree.c

```diff
@@ -775,10 +775,21 @@ __rts_btree_abort_ondisk_kv(WT_SESSION_IMPL *session, WT_REF *ref, WT_ROW *rip,
       upd->type == WT_UPDATE_TOMBSTONE ? "true" : "false",
       __wt_key_string(session, key->data, key->size, S2BT(session)->key_format, key_string));
 
+    /*
+     * It is possible for both RTS and the eviction thread to remove entries from the history store
```

Comment:
I don't understand how this is possible. Doesn't eviction take exclusive access to the page to prevent rts from operating on the same page?

Reply:
The eviction of a table `T`  works on two pages: the page that it's evicting in `T`, and pages with the corresponding keys in the HS. It locks the page that it's evicting, but it updates the HS in a lock-free manner, just like any other update.

Reply:
@kommiharibabu Should rts have already fixed the history store records when it works on the data store if we bring back a version from hs to replace the data store value or we delete the data store value using a globally visible tombstone?

Reply:
This is not actually what's happening. The test that failed ran RTS before connection close, and connection close then runs RTS again. The page that's being evicted from the table was last modified during the first RTS run, and it conflicts with the HS pass of the second RTS run.

Reply:
Still I think the first RTS run should have fixed the relevant hs records and the second rts run should do nothing?

Reply:
You don't need RTS to run into needing to modify the HS during eviction, even though that's what's happening in the test. The problem is that we need to run `__wti_rec_hs_clear_on_tombstone` during reconciliation for correctness, and this is what conflicts with RTS.

It doesn't seem there is currently a way around modifying HS during eviction; Luke P.'s explanation (from Slack): "It is necessary for correctness as the key is therefore deleted. If that key was re-inserted and we didn't delete the corresponding HS entries in theory we could walk them as part of the next search. We'd then be returning values behind a globally deleted point which violates timestamp semantics."

Link: https://mongodb.enterprise.slack.com/archives/G8HA2SRBP/p1717706415993639?thread_ts=1717706223.527089&channel=G8HA2SRBP&message_ts=1717706415.993639

I agree with you though that it would be better if that wasn't necessary, but it seems important the way HS works now.

Reply:
I think I understand it. We have a non-timestamped delete before rts. Then we run rts on the page, it will keep the non-timestamped delete as it is stable. After that, the page is evicted and we need to remove the data from hs.